qcodes/utils/update/qcodes_updater.py

```diff
@@ -0,0 +1,127 @@
+        env_backup_name: Name of the back up environment to be created
+
+    Raises:
+        git.exc.GitError: If working tree is dirty
+    """
+    # QCoDeS source path
+    _source = os.sep.join(qcodes.__file__.split(os.sep)[:-2])
+    # Make sure that git working tree is clean
+    repo = git.Repo(_source)
+    if (inspect.stack()[1].function == 'update_qcodes_installation' and repo.is_dirty()):
+        raise git.exc.GitError
+    elif (inspect.stack()[1].function == 'conda_env_update' and repo.is_dirty()):
+        raise git.exc.GitError
+    elif repo.is_dirty():
+        print("QCoDeS working tree is not clean. Please commit or stash your changes, and try again.")
+    else:
+        _time = str(datetime.date.today()) # Current date.
+        env_backup_name = env_backup_name + '_' + _time
+
+        print(f"Existing {env_name} environment will be backed up as {env_backup_name}...\n")
+        subprocess.run(f'conda create --name {env_backup_name} --clone {env_name}', shell=True)
+
+        # Copy QCoDeS root to qcodes_backup
+        _destination = _source + '_backup' + '_' + _time
+        shutil.copytree(_source, _destination, symlinks=True, ignore=None)
+
+        # Re-install QCoDeS from the back up root
+        subprocess.run(f'activate {env_backup_name} && pip uninstall qcodes && pip install -e {_destination}', shell=True)
+
+        return _source, _destination, env_name, env_backup_name
+
+def conda_env_update(conda_update: bool = True, env_update: bool = True):
+    """
+    Update Conda package manager and QCoDeS Anaconda Python environment
+
+    Args:
+        conda_update: Update Conda pakage manager
+        env_update: Update Anaconda Python environment
+
+    Raises:
+        git.exc.GitError: If working tree is dirty
+    """
+    def execute_update(_source):
+        # Pull QCoDeS master
+        repo = git.Repo(_source)
+        _git = repo.git
+        _git.checkout('master')
+        _git.pull()
+
+        # Update Conda
+        if conda_update:
+            print("Updating Conda...\n")
+            subprocess.run('conda update -n base conda -c defaults', shell=True)
+
+        # Update QCoDeS Anaconda Python environment
+        # Note that Conda will not upgrade a package, even if there is a newer
+        # version, in the case that the package satisfies the requirement specified
+        # by QCoDeS.
+        if env_update:
+            print("Updating QCoDeS environment...\n")
+            subprocess.run('conda env update', shell=True, cwd=_source)
+
+    try:
+        if not inspect.stack()[1].function == 'update_qcodes_installation':
+            # Back up first
+            _source, _destination, env_name, env_backup_name = qcodes_backup()
+            # Then, update
+            execute_update(_source)
+        else:
+            # QCoDeS source path
+            _source = os.sep.join(qcodes.__file__.split(os.sep)[:-2])
+            execute_update(_source)
```

Comment:
Again I really want to discourage functions where the functionality depends on who calls them.